### DISCUSSION TITLE
Good for 3D game development?

### DISCUSSION THREAD

**Discussion by idontplayminecraft** [MonoGame/MonoGame#8496] [General]
Me and a few friends are trying to make an FPS game in the style of ULTRAKILL, DMC and Quake. They said we should use Unity, but since you need a subscription to get rid of the watermark I am looking for alternatives. I heard about MonoGame in a YouTube video. Would it be suited for our game?

**Comment by idontplayminecraft** [comment]
They don’t want to use Godot because it’s “better suited for 2D games!”

**Reply by mandala-dev** [reply]
Since last couple of weeks I've been exploring MonoGame from the 3D point of view, so I can also definitely say - yes.
The results can be seen in a repo for playing with 3D using the latest MonoGame  (https://github.com/LemiBijafra/MonoGame3D) which might help you during the evaluation process.

**Reply by nightblade9** [reply]
Godot 4.x has come a long way in terms of 3D. Whether it's better suited for 2D, I don't know; it supports 3D games pretty well, from what I've seen.

**Comment by Brett208** [comment]
Hello @idontplayminecraft,

Yes, MonoGame can certainly support a first person type game. Another aspect to consider besides licensing is if you are willing to home roll a lot of the features. Some other SDKs/frameworks may offer built in solutions that could shorten your development time if they provide the specific services you need. MonoGame tends to be a more boilerplate solution, which gives a lot of freedom, but will have to develop functionality that may be available out of the box in other frameworks.

Do you plan to use ray casting or a newer technique?

Here is an example of ray casting in MonoGame: https://github.com/u2roka/Heretic

Here is an example closer to what you might see in something like Minecraft (a bit ironic considering your username) called Infiniminer: https://github.com/AristurtleDev/Infiniminer

I've never developed a first person game, so someone stronger in the subject matter might be able to provide a bit better advice.

-Brett

**Reply by idontplayminecraft** [reply]
i see, thanks

**Comment by ghost** [comment]
I'm also interested in 3D.
What classes of MonoGame should I look at specifically for 3D?

**Reply by MonogameGuy** [reply]
The `Model` class is the basis for 3D games.  You will also need a Camera class.  There are several good examples of custom camera classes for XNA.

**Comment by maximm8** [comment]
Yes and NO.  You can definitely do 3D games with monogame, but you will have to program all the aspects of the game, literally everything has to be done from scratch (shaders, physics, UI). I've made a very basic 3D game (Lost Bits) and and it was a lot of work and pain. Another thing, is that not all the features are available on all the platforms, for example multiple render target is not available on IOS.  Overall I would not recommend monogame for 3D, with unity or unreal you will get a lot free stuff, with monogame you get nothing (unless your goal is to build your own engine)

**Comment by rdarnold** [comment]
Honestly I'm not sure that I would recommend it.  I did a 3D game with Monogame, and although I love how it came out, it was really a lot of work.  It was interesting and a super cool learning experience, but in particular learning how to code shaders was a lot.  It's hard when you're like, "I want this sword to look like metal" but there really isn't any way to do that out of the box, you're literally reading tutorials on reflection and light theory and implementing it in HLSL while at the same time reading an HLSL tutorial.  When it works, it's like, holy crap I did that!  I'm really happy I know how to do it NOW but if I were to do it again I might choose a different engine for that style of game even though I love Monogame and almost all my games are done with it.
